### PR TITLE
Lsp scaffold

### PR DESCRIPTION
**Pull request by haifeng-li-at-salesforce**
### What does this PR do?

- LSP client/server basic setup, which includes mocha test framework
- Babel JS parsing with diagnostic rule setup

### CODE REVIEW

File: lsp/server/src/validateDocument.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import { TextDocument } from 'vscode-languageserver-textdocument';
+
+import { Node } from '@babel/types';
+import { DiagnosticProducer } from './diagnostic/DiagnosticProducer';
+import { AdaptersLocalChangeNotAware } from './diagnostic/js/adapters_localChangeNotAware';
+import { parseJs } from './utils/babelUtil';
+
+const jsDiagnosticProducers: DiagnosticProducer<Node>[] = [];
+jsDiagnosticProducers.push(new AdaptersLocalChangeNotAware());
+import { getDocumentSettings } from './server';
```
Comment (sfdctaka) [inline lsp/server/src/validateDocument.ts:18]: nit: `import` should be group together at the top of the file.

**Comment by sfdctaka** [comment]
Curious, do you know if it's a standard practice to put LSP client and server together in an umbrella project(this project I mean, the VSCode extension)? In our case it seems it is fine to do but from project scaffolding point of view it feels unique. I'm wondering if there is a way so we only have a single package.json instead of 3 in the repo.

File: lsp/server/src/validateDocument.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+/**
+ * process the document based extension type.
+ * if html, call html related rules;
+ * if js then parse it using babel, call js related rules
+ * find the gql taggedTemplates, parse the graphql string and call graphql related rules.
+ * @param document the input document to validate.
```
Comment (sfdctaka) [inline lsp/server/src/validateDocument.ts:25]: Since we're JSDoc'ing let's also add `@returns`.

File: lsp/server/src/validateDocument.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+                        break;
+                    }
+
+                    const diagnostics = await producer.validateDocument(
+                        document,
+                        jsNode
+                    );
+                    results.push(...diagnostics);
+                }
+            } catch (e) {}
```
Comment (sfdctaka) [inline lsp/server/src/validateDocument.ts:50]: Why empty catch? If there's an error don't we want it to bubble up if we're not handling it?

File: lsp/server/src/validateDocument.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+                for (const producer of jsDiagnosticProducers) {
+                    if (results.length > setting.maxNumberOfProblems) {
+                        break;
+                    }
+
+                    const diagnostics = await producer.validateDocument(
+                        document,
+                        jsNode
+                    );
+                    results.push(...diagnostics);
```
Comment (sfdctaka) [inline lsp/server/src/validateDocument.ts:48]: Adding spread'ed `diagnostics` in `results` could go over the `maxNumberOfProblems`. Do you need to cull?

File: lsp/server/src/validateDocument.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+export async function validateDocument(
+    document: TextDocument
+): Promise<Diagnostic[]> {
+    const setting = await getDocumentSettings(document.uri);
+
+    const results: Diagnostic[] = [];
+
+    if (document.languageId === 'javascript') {
+        // handles JS rules
+        if (jsDiagnosticProducers.length > 0) {
```
Comment (sfdctaka) [inline lsp/server/src/validateDocument.ts:36]: These `if`s should be combined.

File: lsp/server/src/diagnostic/js/adapters_localChangeNotAware.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        traverse(ast, {
+            Decorator(path) {
+                const expression = path.node.expression;
+                if (isCallExpression(expression)) {
+                    const callee = expression.callee;
+                    if (
+                        callee.type === 'Identifier' &&
+                        callee.name === 'wire'
+                    ) {
+                        if (
```
Comment (sfdctaka) [inline lsp/server/src/diagnostic/js/adapters_localChangeNotAware.ts:66]: Can these `if`s be combined?

**Comment by haifeng-li-at-salesforce** [comment]
Including server/client submodules is the standard practice for LSP. See the MS LSP example https://github.com/salesforce/salesforcedx-vscode-mobile/pull/122.  It provides better modularity and code management.

File: lsp/server/src/validateDocument.ts (hunk 1 above)
Comment (haifeng-li-at-salesforce) [reply lsp/server/src/validateDocument.ts:18]: updated.

File: lsp/server/src/validateDocument.ts (hunk 2 above)
Comment (haifeng-li-at-salesforce) [reply lsp/server/src/validateDocument.ts:25]: updated

File: lsp/server/src/validateDocument.ts (hunk 3 above)
Comment (haifeng-li-at-salesforce) [reply lsp/server/src/validateDocument.ts:50]: This error happens when JS parsing fails, which happens often when developers edit their JS files. It is not practical to keep rendering the parsing error to them since they are in middle of the editing, plug other tool kits can also give them the hint.

File: lsp/server/src/validateDocument.ts (hunk 4 above)
Comment (haifeng-li-at-salesforce) [reply lsp/server/src/validateDocument.ts:48]: Nice finding! Code updated and `diagnostics` is culled

File: lsp/server/src/validateDocument.ts (hunk 5 above)
Comment (haifeng-li-at-salesforce) [reply lsp/server/src/validateDocument.ts:36]: JS handling is separated.  Graphql and html handling can be added later

File: lsp/server/src/diagnostic/js/adapters_localChangeNotAware.ts (hunk 6 above)
Comment (haifeng-li-at-salesforce) [reply lsp/server/src/diagnostic/js/adapters_localChangeNotAware.ts:66]: Good finding! Updated.

File: lsp/server/src/validateDocument.ts (hunk 3 above)
Comment (sfdctaka) [reply lsp/server/src/validateDocument.ts]: Although I still find it an odd way to early exit the validation while still typing/editing the code I also can't think of any way better to do this. Thanks for adding the comment. As long as we can tell why there's an empty catch I can live with this :)

File: lsp/server/src/test/diagnostic/js/adapters_localChangeNotAware.spec.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+import { AdaptersLocalChangeNotAware } from '../../../diagnostic/js/adapters_localChangeNotAware';
+import { parseJs } from '../../../utils/babelUtil';
+import { TextDocument } from 'vscode-languageserver-textdocument';
+import { text } from 'stream/consumers';
+
+const relatedRecordsJS = `
+import { LightningElement, wire } from "lwc";
+import { getRelatedListRecords } from "lightning/uiRelatedListApi";
+
+export default class RelatedListRecords extends LightningElement {
```
Comment (haifeng-li-at-salesforce) [inline lsp/server/src/test/diagnostic/js/adapters_localChangeNotAware.spec.ts:20]: E2E test will be added in separate PR

File: lsp/client/package.json [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+        "vscode": "1.77.0"
+    },
+    "dependencies": {
+        "vscode-languageclient": "^9.0.1"
+    },
+    "devDependencies": {
+        "@types/vscode": "1.77.0",
+        "@vscode/test-electron": "^2.3.9"
+    }
+}
```
Comment (sfdctaka) [inline lsp/client/package.json:22]: Can we update package.json for client and server with appropriate values.

File: lsp/client/package.json [hunk 9]
```diff
@@ -0,0 +1,22 @@
+{
+    "name": "lsp-sample-client",
```
Comment (ben-zhang-at-salesforce) [inline lsp/client/package.json:2]: better name

File: lsp/client/package.json [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+    "description": "VSCode part of a language server",
+    "author": "Microsoft Corporation",
```
Comment (ben-zhang-at-salesforce) [inline lsp/client/package.json:4]: remove author? root project already have it

File: lsp/client/package.json [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+    "license": "MIT",
+    "version": "0.0.1",
+    "publisher": "vscode",
+    "repository": {
```
Comment (ben-zhang-at-salesforce) [inline lsp/client/package.json:8]: remove it?

File: lsp/client/src/extension.ts [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+        ],
+        synchronize: {
+            // Notify the server about file changes to '.clientrc files contained in the workspace
+            fileEvents: workspace.createFileSystemWatcher('**/.clientrc')
+        }
+    };
+
+    // Create the language client and start the client.
+    client = new LanguageClient(
+        'lsp Mobile',
```
Comment (ben-zhang-at-salesforce) [inline lsp/client/src/extension.ts:51]:
```suggestion
        'LSP Mobile',
```

File: lsp/server/package.json [hunk 13]
```diff
@@ -0,0 +1,31 @@
+{
+    "name": "lsp-sample-server",
```
Comment (ben-zhang-at-salesforce) [inline lsp/server/package.json:2]: same in this file,  polish/clean up like author, name, description, repository ..?

Comment (haifeng-li-at-salesforce) [reply lsp/server/package.json:2]: Good catch.

File: lsp/client/package.json (hunk 8 above)
Comment (haifeng-li-at-salesforce) [reply lsp/client/package.json:22]: Change it to `mobile-lsp-server` and `mobile-lsp-client`

File: lsp/client/package.json (hunk 10 above)
Comment (haifeng-li-at-salesforce) [reply lsp/client/package.json:4]: done

File: lsp/client/package.json (hunk 11 above)
Comment (haifeng-li-at-salesforce) [reply lsp/client/package.json:8]: done

File: lsp/server/src/diagnostic/js/adapters_localChangeNotAware.ts [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+    /**
+     * Find @wire adapter call which called in the local change not aware adapters. For example: 
+        export default class RelatedListRecords extends LightningElement {
+            ...
+            @wire(getRelatedListRecords, 
+            ...
+        }
+     * @param astNode root node to search
+     * @param adapterNames adapter which are not able to reflect the local change.
+     * @returns node with adapter name
```
Comment (ben-zhang-at-salesforce) [inline lsp/server/src/diagnostic/js/adapters_localChangeNotAware.ts:57]:
```suggestion
     * @returns nodes with adapter name
```

File: lsp/client/src/extension.ts (hunk 12 above)
Comment (haifeng-li-at-salesforce) [reply lsp/client/src/extension.ts:51]: DONE

Comment (sfdctaka) [inline lsp/client/src/extension.ts:51]: nit: Why mixed case 'lsp' in all lower case with 'Mobile' having only first character in caps? Is it legal and correct to do 'LSP Mobile Client' like the second parameter?

File: lsp/server/src/diagnostic/js/adapters_localChangeNotAware.ts (hunk 14 above)
Comment (haifeng-li-at-salesforce) [reply lsp/server/src/diagnostic/js/adapters_localChangeNotAware.ts:57]: done

**Review by ben-zhang-at-salesforce — APPROVED** [review]
looks good to me.  great job!

**Review by sfdctaka — APPROVED** [review]
LGTM
